### ISSUE TITLE
Handling of misformatted JSON

### ISSUE THREAD

**Issue by ml31415** [jcrist/msgspec#424]
### Description

I'm working on a client library for an API product of some multibillion dollar company, that seems to use some proprietary server software, probably from another multibillion dollar company. Neither of them seem to give a .... have the necessary diligence, to make their products fully JSON specification compatible. Or some other top secret business reasons, like not enough budget for some programmer hours, to fix it.

Anyways, what I want to say, it's something that probably won't change and is something the client side will have to live and deal with. And it's probably something, that shouldn't be too uncommon to be found out there in the "wild".

What you irregularly find in these API results is `"null"` instead of `null`, `"123"` instead of `123`, `"123.45"` instead of `123.45`and so on. Of course not consistently, some response fields have the erroneous quotation always, some occasionally, and some are encoded correctly. So far the best approach I found to deal with it, is this:

```python
import msgspec

class IntStr(int):
    pass
    
class FloatStr(float):
    pass


def encode_quoted(obj):
    if isinstance(obj, int):
        return int(obj)
    if isinstance(obj, float):
        return float(obj)
    raise TypeError(f"Unencodable type: {type(obj).__name__}: {obj}")


def decode_quoted(type_, obj):
    if type_ is IntStr:
        if isinstance(obj, int):
            return IntStr(obj)
        if obj == "null":
            return IntStr(0)
        return IntStr(obj.strip("'\" "))
    if type_ is FloatStr:
        if isinstance(obj, (float, int)):
            return FloatStr(obj)
        if obj == "null":
            return FloatStr("nan")
        return FloatStr(obj.strip("'\" "))
    raise TypeError(f"Undecodable type: {type(obj).__name__}: {obj}. Expected type: {type_.__name__}")
```

Now, unfortunately this only partly does, what it's supposed to do. I can declare `IntStr` fields, and they get parsed to an `IntStr` object. Ideally, I'd like to have them plain `int`, but this doesn't work. If I try to return `int` instead of `IntStr` from the decode hook, I get a `TypeError`

Also it has issues with optional values:

```python
>>> msgspec.json.decode(b'"123"', type=IntStr, dec_hook=decode_quoted)
123
>>> msgspec.json.decode(b'"null"', type=Optional[IntStr], dec_hook=decode_quoted)
0
```
Ideally, I'd like to have `None` here, but again, I get a TypeError, if I try to return some other type than `IntStr`.

So far I also failed to define a `Union` as a custom type for the decode hook. If there is any possibility, that would already be a big help.

I'm not sure what I should suggest as a best solution for this. Maybe another keyword to `msgspec.field` and `msgspec.Meta` like `attempt_unquote=`? Or the possibility of `decode_hook` to return other types, than the ones that were specified in the type definition?

**Comment by ml31415** [comment]
After reading through all previously documented issues, that would require some extended and improved hook mechanism ( #289 #106 #349 #336 #260 #248 ), I'd like to suggest the following API change, that I guess should solve all presented issues:

- `msgspec.Meta` gets two extra keywords, `decoder` and `encoder`
- `msgspec.field` gets the same two keywords
- `decoder` is a simple function, that if defined will be fed with the parsed JSON standard data of type `int  | float | str | None` and needs to return anything of the specified annotated type or the field annotation.
- This annotated type can also be a Union of currently incompatible types, and it will be the decoders job to decide which one to pick. If it fails to return a correct type, a type error should be thrown just as happens right now,
- `encoder` will also be a simple function, that takes whatever the user throws into it, and needs to return anything of the standard JSON types `int  | float | str | None`.
- It will be the users responsibility, that encoder and decoder fit together and can distinguish between so far indistinguishable types, subclasses and so on.
- Providing a decoding hook wouldn't be necessary anymore for those types, as all information how to deal with them is contained in the type or field annotation
- Nevertheless, the current hook mechanism could stay as is, not touching backwards compatibility, but maybe deprecated later on
- No separate hooks registration and so on would be required

It would also be possible, to have a Union of such special annotated classes. In that case, if there is not explicitly a decoder defined, I'd suggest to try the decoders in sequence, until one returns a result, TypeError if no one works.

So to give an example: With my use-case above, I'd define 

```python

def none_decoder(val):
    if val == "null" or val is None:
        return None
    raise TypeError


IntStr = Annotated[int, msgspec.Meta(title="IntStr", decoder=int)]
FloatStr = Annotated[float, msgspec.Meta(title="FloatStr", decoder=float)]
NoneStr = Annotated[None, msgspec.Meta(title="NoneStr", decoder=none_decoder)]

QuotedOptionalNumber = Annotated[Union[Intstr,FloatStr,NoneStr], msgspec.Meta(title="OuotedOptionalNumber")]
```

Maybe I'd also just go with:

```python
def unquote(val:
    try:
        return int(val)
    except ValueError:
        try:
            return float(val)
        except ValueError(val):
            pass
    if val == "null":
        return None
    raise TypeError


ErrQuoted = Annotated[int| float | None, msgspec.Meta(title="ErrQuoted", decoder=unquote)]
````
Or another example for #349 :

```python
def decimal_decode(val):
    try:
        return decimal.Decimal(val)
    except decimal.InvalidOperation
        raise TypeError
        

Decimal = Annotated[decimal.Decimal, msgspec.Meta(title="Decimal", decoder=decimal_decode)]
DecimalOrStr = Annotated[Union[Decimal,str], title="DecimalOrStr")]

# another alternative

def decimal_or_str(val):
    try:
        return decimal.Decimal(val)
    except decimal.InvalidOperation
        return str(val
    
DecimalOrStr = Annotated[Union[decimal.Decimal,str], title="DecimalOrStr", decoder=decimal_or_str)]
```

I hope I didn't forget anything vital and I bet there are still a couple of corners, that need some extra thought, Though, I guess that might

**Comment by jcrist** [comment]
Luckily all of the cases above match our current coercion rules for the `str_values=True` option to `msgspec.from_builtins`. I think what I'd want to do here is:

- Rename the `from_builtins` option from `str_values=True` to `coerce=True` (or maybe `strict=False`? not sure on a good name, thoughts welcome here)
- Add this option to all builtin decoders. When enabled, we'd support more coercion of types to try and fit the provided schema.

In your case then you could continue to use the standard type definitions:

```python
msgspec.json.decode(msg, type=YourType, coerce=True)
```

**Comment by jcrist** [comment]
This has been resolved in #434. When decoding, you can pass `strict=False` to either `decode` or when creating a `Decoder` to enable a wider set of coercion rules.

I expect to cut a release in a week or so, but if you want to try this out before then you can [install from github](https://jcristharif.com/msgspec/install.html#installing-from-github).

**Comment by ml31415** [comment]
Thanks for the quick fix!

**Comment by jcrist** [comment]
This has been released as part of [0.16.0](https://github.com/jcrist/msgspec/releases/tag/0.16.0), available on pypi and conda-forge now.
